README
README, xc - aXis Controller, Readme file

1. Introduction

5. Configuration file

5.1. Information

5.2. Macros
I recommend you use underline before macro name. It's usefull to quickly identify a macro.

5.3. Keyboard map
╔════════════════╦═══════╦═════════════════════════════════════════════════════╗
║    KeyASCII    ║ ASCII ║                    Common Name                      ║
╠════════════════╬═══════╬═════════════════════════════════════════════════════╣
║ K_BACKSPACE    ║ \b    ║ backspace                                           ║
║ K_TAB          ║ \t    ║ tab                                                 ║
║ K_CLEAR        ║       ║ clear                                               ║
║ K_RETURN       ║ \r    ║ return                                              ║
║ K_PAUSE        ║       ║ pause                                               ║
║ K_ESCAPE       ║ ^[    ║ escape                                              ║
║ K_SPACE        ║       ║ space                                               ║
║ K_EXCLAIM      ║ !     ║ exclaim                                             ║
║ K_QUOTEDBL     ║ "     ║ quotedbl                                            ║
║ K_HASH         ║ #     ║ hash                                                ║
║ K_DOLLAR       ║ $     ║ dollar                                              ║
║ K_AMPERSAND    ║ &     ║ ampersand                                           ║
║ K_QUOTE        ║       ║ quote                                               ║
║ K_LEFTPAREN    ║ (     ║ left parenthesis                                    ║
║ K_RIGHTPAREN   ║ )     ║ right parenthesis                                   ║
║ K_ASTERISK     ║ *     ║ asterisk                                            ║
║ K_PLUS         ║ +     ║ plus sign                                           ║
║ K_COMMA        ║ ,     ║ comma                                               ║
║ K_MINUS        ║ -     ║ minus sign                                          ║
║ K_PERIOD       ║ .     ║ period                                              ║
║ K_SLASH        ║ /     ║ forward slash                                       ║
║ K_0            ║ 0     ║ 0                                                   ║
║ K_1            ║ 1     ║ 1                                                   ║
║ K_2            ║ 2     ║ 2                                                   ║
║ K_3            ║ 3     ║ 3                                                   ║
║ K_4            ║ 4     ║ 4                                                   ║
║ K_5            ║ 5     ║ 5                                                   ║
║ K_6            ║ 6     ║ 6                                                   ║
║ K_7            ║ 7     ║ 7                                                   ║
║ K_8            ║ 8     ║ 8                                                   ║
║ K_9            ║ 9     ║ 9                                                   ║
║ K_COLON        ║ :     ║ colon                                               ║
║ K_SEMICOLON    ║ ;     ║ semicolon                                           ║
║ K_LESS         ║ <     ║ less-than sign                                      ║
║ K_EQUALS       ║ =     ║ equals sign                                         ║
║ K_GREATER      ║ >     ║ greater-than sign                                   ║
║ K_QUESTION     ║ ?     ║ question mark                                       ║
║ K_AT           ║ @     ║ at                                                  ║
║ K_LEFTBRACKET  ║ [     ║ left bracket                                        ║
║ K_BACKSLASH    ║ \     ║ backslash                                           ║
║ K_RIGHTBRACKET ║ ]     ║ right bracket                                       ║
║ K_CARET        ║ ^     ║ caret                                               ║
║ K_UNDERSCORE   ║ _     ║ underscore                                          ║
║ K_BACKQUOTE    ║ `     ║ grave                                               ║
║ K_a            ║ a     ║ a                                                   ║
║ K_b            ║ b     ║ b                                                   ║
║ K_c            ║ c     ║ c                                                   ║
║ K_d            ║ d     ║ d                                                   ║
║ K_e            ║ e     ║ e                                                   ║
║ K_f            ║ f     ║ f                                                   ║
║ K_g            ║ g     ║ g                                                   ║
║ K_h            ║ h     ║ h                                                   ║
║ K_i            ║ i     ║ i                                                   ║
║ K_j            ║ j     ║ j                                                   ║
║ K_k            ║ k     ║ k                                                   ║
║ K_l            ║ l     ║ l                                                   ║
║ K_m            ║ m     ║ m                                                   ║
║ K_n            ║ n     ║ n                                                   ║
║ K_o            ║ o     ║ o                                                   ║
║ K_p            ║ p     ║ p                                                   ║
║ K_q            ║ q     ║ q                                                   ║
║ K_r            ║ r     ║ r                                                   ║
║ K_s            ║ s     ║ s                                                   ║
║ K_t            ║ t     ║ t                                                   ║
║ K_u            ║ u     ║ u                                                   ║
║ K_v            ║ v     ║ v                                                   ║
║ K_w            ║ w     ║ w                                                   ║
║ K_x            ║ x     ║ x                                                   ║
║ K_y            ║ y     ║ y                                                   ║
║ K_z            ║ z     ║ z                                                   ║
║ K_DELETE       ║       ║ delete                                              ║
║ K_KP0          ║       ║ keypad 0                                            ║
║ K_KP1          ║       ║ keypad 1                                            ║
║ K_KP2          ║       ║ keypad 2                                            ║
║ K_KP3          ║       ║ keypad 3                                            ║
║ K_KP4          ║       ║ keypad 4                                            ║
║ K_KP5          ║       ║ keypad 5                                            ║
║ K_KP6          ║       ║ keypad 6                                            ║
║ K_KP7          ║       ║ keypad 7                                            ║
║ K_KP8          ║       ║ keypad 8                                            ║
║ K_KP9          ║       ║ keypad 9                                            ║
║ K_KP_PERIOD    ║ .     ║ keypad period                                       ║
║ K_KP_DIVIDE    ║ /     ║ keypad divide                                       ║
║ K_KP_MULTIPLY  ║ *     ║ keypad multiply                                     ║
║ K_KP_MINUS     ║ -     ║ keypad minus                                        ║
║ K_KP_PLUS      ║ +     ║ keypad plus                                         ║
║ K_KP_ENTER     ║ \r    ║ keypad enter                                        ║
║ K_KP_EQUALS    ║ =     ║ keypad equals                                       ║
║ K_UP           ║       ║ up arrow                                            ║
║ K_DOWN         ║       ║ down arrow                                          ║
║ K_RIGHT        ║       ║ right arrow                                         ║
║ K_LEFT         ║       ║ left arrow                                          ║
║ K_INSERT       ║       ║ insert                                              ║
║ K_HOME         ║       ║ home                                                ║
║ K_END          ║       ║ end                                                 ║
║ K_PAGEUP       ║       ║ page up                                             ║
║ K_PAGEDOWN     ║       ║ page down                                           ║
║ K_F1           ║       ║ F1                                                  ║
║ K_F2           ║       ║ F2                                                  ║
║ K_F3           ║       ║ F3                                                  ║
║ K_F4           ║       ║ F4                                                  ║
║ K_F5           ║       ║ F5                                                  ║
║ K_F6           ║       ║ F6                                                  ║
║ K_F7           ║       ║ F7                                                  ║
║ K_F8           ║       ║ F8                                                  ║
║ K_F9           ║       ║ F9                                                  ║
║ K_F10          ║       ║ F10                                                 ║
║ K_F11          ║       ║ F11                                                 ║
║ K_F12          ║       ║ F12                                                 ║
║ K_F13          ║       ║ F13                                                 ║
║ K_F14          ║       ║ F14                                                 ║
║ K_F15          ║       ║ F15                                                 ║
║ K_NUMLOCK      ║       ║ numlock                                             ║
║ K_CAPSLOCK     ║       ║ capslock                                            ║
║ K_SCROLLOCK    ║       ║ scrollock                                           ║
║ K_RSHIFT       ║       ║ right shift                                         ║
║ K_LSHIFT       ║       ║ left shift                                          ║
║ K_RCTRL        ║       ║ right control                                       ║
║ K_LCTRL        ║       ║ left control                                        ║
║ K_RALT         ║       ║ right alt                                           ║
║ K_LALT         ║       ║ left alt                                            ║
║ K_RMETA        ║       ║ right meta                                          ║
║ K_LMETA        ║       ║ left meta                                           ║
║ K_LSUPER       ║       ║ left Windows key                                    ║
║ K_RSUPER       ║       ║ right Windows key                                   ║
║ K_MODE         ║       ║ mode shift                                          ║
║ K_HELP         ║       ║ help                                                ║
║ K_PRINT        ║       ║ print screen                                        ║
║ K_SYSREQ       ║       ║ sysrq                                               ║
║ K_BREAK        ║       ║ break                                               ║
║ K_MENU         ║       ║ menu                                                ║
║ K_POWER        ║       ║ power                                               ║
║ K_EURO         ║       ║ Euro                                                ║
╚════════════════╩═══════╩═════════════════════════════════════════════════════╝


sudo apt-get install python-serial
sudo apt-get install python-termcolor 
sudo apt-get install python-sh
sudo apt-get install python-pygame
sudo apt-get install python-psutil


CHANGELOG
CHANGELOG, xc - aXis Controller, Change log file

2017-07-27
        * Version: 0.48b
        * Improvement: Added memory information to GUI startup messages.

2017-07-20
        * Version: 0.47b
        * Bug fix: Fan speed correction.

2017-07-18
        * Improvement: Added lots of documentation.

2017-06-01
        * New feature: Added platform machine detection.
        * Improvement: Added status led blink when xc is running.

2017-06-28
        * New feature: Added Status LED control.

2017-06-26
        * New feature: Added fan control and fan speed sensor.

2017-06-13
        * Improvement: Detailed control start messages on GUI.

2017-06-04
        * Bug fix: Fixed DevTools() program execution check.
        * New feature: Added comment parser to caracter '()' in CommandParser().

2017-06-02
        * New feature: Added comment parser to caracter ';' in CommandParser().

2017-05-22
        * New feature: Keyboard and mouse detection.

2017-05-19
        * New feature: Joystick buttons support.

2017-05-17
        * Bug fix: Correction applied to joystick precision.

2017-05-12
        * New feature: Joystick support.

2017-05-11
        * New feature: Mouse support.

2017-05-10
        * New feature: Automatically build object from JSON definition.
        * New feature: Keyboard support.

2017-05-08
        * Improvement: Check for TrueType fonts on start up.

2017-04-30
        * Improvement: list command can now show only connected devices using
          -c or --connected options. Type "xc list -h" for more information.
        * New feature: PWA interface.
        * Improvement: More documentation.

2017-04-03
        * Bug fix: Change escape caracter on xc (shell script) starter program.

2017-03-06
        * New feature: Now list command can force check on a specific interface.

2017-03-05
        * Improvement: Removed unused code. 

2017-02-28
        * Improvement: G-code runner next step is trigged to start when 'ok'
          string is received.

2017-02-27
        * Bug fix: Auto detection was overlapping explicit device ID definition.
        * Bug fix: command list is now too fast.

2017-02-26
        * New feature: Device auto connection.
        * Improvement: Check external apps before execution.
        * Improvement: Optimized and removed repeated code blocks.

2017-02-24
        * New feature: Device auto detection feature.

2017-02-21
        * Strategic: Some changes to support future migration to Python 3.
        * Improvement: Check files before open.
        * New feature: Display FPS (Frames Per Second) information on GUI.

2017-02-20
        * New feature: Class Gui.

2017-02-19
        * Improvement: Moved message_list() funtionalities to 
          UserArgumentParser() class list() method.

2017-02-18
        * Improvement: Migrated DevTools class to devtools package.

2017-02-17
        * New feature: Network interface definition on terminal option.

2017-02-15
        * Bug fix: Help messages and user input arguments and options.
        * New feature: Run programs on CLI.
        * Improvement: Argparse support.

2017-02-12
        * New feature: Connection status is available to list command.

2017-02-07
        * New feature: Added Echo class, to control message level
                       (error, warning, information, code).
        * Improvement: Help messages optimization.

2017-02-03
        * Strategic: Project renamed to xC.

2016-12-11
        * Improvement: Accepting IP address or host name to network functions.

2016-05-26
        * Bug fix: Fixed parameters used on miniterm.py to run on Ubuntu 16.04.

2016-04-12
        * Improvement: Code optionazation, separated into some package files.

2015-10-18
        * Bug fix: list option fixed.

2015-09-11
        * Improvement: Availability libraries check before import.

2015-08-19
        * Improvement: JSON device file optimization and organization.

2015-08-08
        * Improvement: CLI arguments now can by inputed using any order.
        * Improvement: Device check before run some actions.

2015-07-19
        * Improvement: Default command line option changed to run.

2015-07-17
        * Improvement: Converted original Bash code to Python.
                       All features ported.
        * Improvement: Terminal is now based on miniterm.py.

2015-02-06
        * Project renamed to xC.

2014-08-18
        * New feature: Added command option "run"

2014-02-07
        * Bug fix: Minor bug fix.

2014-02-06
        * Initial version written in Bash.


xc.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
"""
xc.py

Copyright (c) 2014-2017 Márcio Pessoa <marcio.pessoa@sciemon.com>

Author: Marcio Pessoa <marcio@pessoa.eti.br>
Contributors: none

Change log: Check CHANGELOG file.

"""

try:
    # Ubuntu default modules
    import sys
    import argparse
    # import getopt
    import os.path
    import time
    # Ubuntu manually installed modules
    import serial
    # Myself modules
    from command_parser import CommandParser
    from device_properties import DeviceProperties
    from devtools import DevTools
    from echo import verbose, level, \
        echo, echoln, erro, erroln, warn, warnln, info, infoln, code, codeln
except ImportError as err:
    print("Could not load module. " + str(err))
    sys.exit(True)


class UserArgumentParser():
    """
    https://docs.python.org/2/library/argparse.html
    http://chase-seibert.github.io/blog/
    """

    # Set default verbosity level
    verbose(1)  # Error level

    def __init__(self):
        self.program_name = "xc"
        self.program_version = "0.48b"
        self.program_date = "2017-07-27"
        self.program_description = "xC - aXes Controller"
        self.program_copyright = "Copyright (c) 2014-2017 Marcio Pessoa"
        self.program_license = "undefined. There is NO WARRANTY."
        self.program_website = "http://pessoa.eti.br/"
        self.program_contact = "Marcio Pessoa <marcio.pessoa@sciemon.com>"
        self.id = None
        self.interface = None
        self.config_file = os.path.join(os.getenv('HOME', ''), '.devices.json')
        header = ('xc <command> [<args>]\n\n' +
                  'commands:\n' +
                  '  gui            graphical user interface\n' +
                  '  cli            command line interface\n' +
                  '  terminal       connect to device terminal\n' +
                  '  verify         check firmware code sintax\n' +
                  '  upload         upload firmware to device\n' +
                  '  list           list devices\n\n')
        footer = (self.program_copyright + '\n' +
                  'License: ' + self.program_license + '\n' +
                  'Website: ' + self.program_website + '\n' +
                  'Contact: ' + self.program_contact + '\n')
        examples = ('examples:\n' +
                    '  xc list\n' +
                    '  xc verify --id x6 --verbosity=3\n' +
                    '  xc upload -i envmon --interface=network\n' +
                    '  xc gui --id x2\n' +
                    '  xc cli -i x6 -p file.gcode -r\n')
        version = (self.program_name + " " + self.program_version + " (" +
                   self.program_date + ")" + '' + '\n' + footer)
        epilog = (examples + '\n' + footer)
        parser = argparse.ArgumentParser(
            prog=self.program_name,
            formatter_class=argparse.RawDescriptionHelpFormatter,
            epilog=epilog,
            add_help=True,
            usage=header)
        parser.add_argument('command', help='command to run')
        parser.add_argument('-v', '--version', action='version',
                            version=version,
                            help='show version information and exit')
        if len(sys.argv) < 2:
            self.gui()
            sys.exit(False)
        args = parser.parse_args(sys.argv[1:2])
        if not hasattr(self, args.command):
            echoln('Unrecognized command')
            parser.print_help()
            sys.exit(True)
        getattr(self, args.command)()

    def cli(self):
        parser = argparse.ArgumentParser(
            prog=self.program_name + ' cli',
            description='connect to device terminal')
        parser.add_argument(
            '-i', '--id',
            help='device ID')
        parser.add_argument(
            '--interface',
            default='serial',
            choices=['serial', 'network'],
            help='communication interface (default: serial)')
        parser.add_argument(
            '-l', '--log', metavar='file',
            help='log file')
        parser.add_argument(
            '-o', '--out', metavar='file',
            help='output file')
        parser.add_argument(
            '-p', '--program', metavar='file',
            help='load G-code file')
        parser.add_argument(
            '-r', '--run', action="store_true",
            help='run program')
        parser.add_argument(
            '--verbosity', type=int,
            default=1,
            choices=[0, 1, 2, 3, 4],
            help='verbose mode, options: ' +
                 '0 Quiet, 1 Errors (default), 2 Warnings, 3 Info, 4 Code')
        args = parser.parse_args(sys.argv[2:])
        verbose(args.verbosity)
        # Start CLI
        from cli import Cli
        if self.__connection(args.id, args.interface):
            sys.exit(True)
        # Start G-code parser
        gcode = CommandParser(self.ser)
        if gcode.load(args.program):
            erroln("Fail reading file: " + args.program)
            sys.exit(True)
        gcode.run()
        sys.exit(False)

    def gui(self):
        parser = argparse.ArgumentParser(
            prog=self.program_name + ' gui',
            description='graphical user interface')
        parser.add_argument(
            '-f', '--fullscreen', action="store_true",
            help='full screen')
        parser.add_argument(
            '-i', '--id',
            help='device ID')
        parser.add_argument(
            '--interface',
            default='serial',
            choices=['serial', 'network'],
            help='communication interface (default: serial)')
        parser.add_argument(
            '-l', '--log', metavar='file',
            help='log file')
        parser.add_argument(
            '-o', '--out', metavar='file',
            help='output file')
        parser.add_argument(
            '-p', '--program', metavar='file',
            help='load G-code file')
        parser.add_argument(
            '-r', '--run', action="store_true",
            help='run program')
        parser.add_argument(
            '--verbosity', type=int,
            default=1,
            choices=[0, 1, 2, 3, 4],
            help='verbose mode, options: ' +
                 '0 Quiet, 1 Errors (default), 2 Warnings, 3 Info, 4 Code')
        args = parser.parse_args(sys.argv[2:])
        verbose(args.verbosity)
        # Start GUI
        from gui import Gui
        gui = Gui()
        gui.config_file = self.config_file
        gui.id = args.id
        gui.interface = args.interface
        gui.start(args.fullscreen)
        gui.run()
        sys.exit(False)

    def terminal(self):
        parser = argparse.ArgumentParser(
            prog=self.program_name + ' terminal',
            description='command line interface')
        parser.add_argument(
            '-i', '--id',
            help='device ID')
        parser.add_argument(
            '--interface',
            default='serial',
            choices=['serial', 'network'],
            help='communication interface (default: serial)')
        parser.add_argument(
            '--verbosity', type=int,
            default=1,
            choices=[0, 1, 2, 3, 4],
            help='verbose mode, options: ' +
                 '0 Quiet, 1 Errors (default), 2 Warnings, 3 Info, 4 Code')
        args = parser.parse_args(sys.argv[2:])
        verbose(args.verbosity)
        self.__dev_tools(args.id, args.interface)
        self.project.terminal()

    def verify(self):
        parser = argparse.ArgumentParser(
            prog=self.program_name + ' verify',
            description='check firmware code')
        parser.add_argument(
            '-i', '--id',
            help='device ID')
        parser.add_argument(
            '--verbosity', type=int,
            default=1,
            choices=[0, 1, 2, 3, 4],
            help='verbose mode, options: ' +
                 '0 Quiet, 1 Errors (default), 2 Warnings, 3 Info, 4 Code')
        args = parser.parse_args(sys.argv[2:])
        verbose(args.verbosity)
        self.__dev_tools(args.id)
        self.project.verify()

    def upload(self):
        parser = argparse.ArgumentParser(
            prog=self.program_name + ' upload',
            description='upload firmware to device')
        parser.add_argument(
            '-i', '--id',
            help='device ID')
        parser.add_argument(
            '--interface',
            default='serial',
            choices=['serial', 'network'],
            help='communication interface (default: serial)')
        parser.add_argument(
            '--verbosity', type=int,
            default=1,
            choices=[0, 1, 2, 3, 4],
            help='verbose mode, options: ' +
                 '0 Quiet, 1 Errors (default), 2 Warnings, 3 Info, 4 Code')
        args = parser.parse_args(sys.argv[2:])
        verbose(args.verbosity)
        self.__dev_tools(args.id, args.interface)
        self.project.upload()

    def list(self):
        parser = argparse.ArgumentParser(
            prog=self.program_name + ' list',
            description='list devices')
        parser.add_argument(
            '-c', '--connected', action="store_true",
            help='show only connected devices')
        parser.add_argument(
            '--interface',
            default='all',
            choices=['all', 'serial', 'network'],
            help='communication interface (default: all)')
        parser.add_argument(
            '--verbosity', type=int,
            default=1,
            choices=[0, 1, 2, 3, 4],
            help='verbose mode, options: ' +
                 '0 Quiet (return number of devices), 1 IDs (default),' +
                 ' 2 Names, 3 Status, 4 Description')
        args = parser.parse_args(sys.argv[2:])
        device = DeviceProperties(self.config_file)
        if args.verbosity >= 2:
            echo(' Id\tName\tMark')
        if args.verbosity >= 3:
            echo('\tLink')
        if args.verbosity >= 4:
            echo('\t Description')
        if args.verbosity > 1:
            echoln('')
        if args.verbosity > 1:
            for i in range(args.verbosity):
                echo("-------------")
            echoln('')
        c = 0
        for id in device.list():
            device.select(id)
            if args.verbosity >= 3 or args.connected:
                interface = '-'
                if args.interface == 'serial' or args.interface == 'all':
                    if device.is_serial_connected():
                        interface = "Serial"
                if args.interface == 'network' or args.interface == 'all':
                    if device.is_network_connected():
                        interface = "Network"
            if args.connected and interface == '-':
                continue
            if args.verbosity >= 1:
                echo(id)
            if args.verbosity >= 2:
                echo("\t" +
                     device.system_plat + "\t" +
                     device.system_mark)
            if args.verbosity >= 3:
                echo('\t' + interface)
            if args.verbosity >= 4:
                echo('\t ' + device.system_desc)
            if args.verbosity > 0:
                echoln('')
            c += 1
        sys.exit(c)

    def __connection(self, id=None, interface=None):
        # Load device configuration file
        self.device = DeviceProperties(self.config_file)
        # Select device
        if id:
            self.id = id
            self.device.select(self.id)
        else:
            if self.device.select_auto():
                self.device.presentation()
            else:
                return False
        # Connect to device
        if interface:
            self.interface = interface
            infoln("Connecting...")
            if self.interface == 'serial':
                if self.device.is_serial_connected():
                    self.ser = serial.Serial(
                        self.device.comm_serial_path,
                        self.device.comm_serial_speed,
                        timeout=1)
                else:
                    erroln('Device is not connected.')
                    infoln('Try again after connecting the device: ' +
                           str(self.device.comm_serial_path))
                    return True
            elif self.interface == 'network':
                if self.device.comm_network_address is None:
                    erroln("Interface not configured for this device.")
                    return True
                if not self.device.is_network_connected():
                    erroln('Device is not reacheable: ' +
                           str(self.device.comm_network_address))
                    return True

    def __dev_tools(self, id=None, interface=None):
        if id:
            self.id = id
        if interface:
            self.interface = interface
        infoln('Started at: \'' + time.strftime('%Y-%m-%d %H:%M:%S') + '\'')
        # Load device configuration file
        self.device = DeviceProperties(self.config_file)
        if self.id is None:
            self.id = self.device.select_auto()
        else:
            self.device.select(self.id)
        self.device.presentation()
        self.project = DevTools(self.device.id,
                                self.device.system_plat,
                                self.device.system_mark,
                                self.device.system_desc,
                                self.device.system_arch,
                                self.device.system_path,
                                self.device.system_logs,
                                self.device.comm_serial_path,
                                self.device.comm_serial_speed,
                                self.device.comm_network_address,
                                self.interface)
        if self.interface:
            infoln("Connecting...")
            if self.interface == 'serial':
                if not self.device.is_serial_connected():
                    erroln('Serial device is not connected.')
                    infoln('Try again after connecting the device: ' +
                           str(self.device.comm_serial_path))
                    sys.exit(True)
            elif self.interface == 'network':
                if self.device.comm_network_address is None:
                    erroln("Network interface not configured for this device.")
                    sys.exit(True)
                if not self.device.is_network_connected():
                    erroln('Network device is not reacheable.')
                    infoln('Check device connectivity: ' +
                           str(self.device.comm_network_address))
                    sys.exit(True)


def main():
    UserArgumentParser()


if __name__ == '__main__':
    main()
